File: iOS-ObjC/watf-bank/PageController/History.m
#import "History.h"

@interface History()
@end

@implementation History
- (void)viewDidLoad
{
    [super viewDidLoad];
    [_table setDelegate:self];
    [_table setDataSource:self];
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        self.fromAccountData = (self.fromAccountData==nil) ? [[NSMutableArray alloc] init] : self.fromAccountData;
        self.toAccountData = (self.toAccountData==nil) ? [[NSMutableArray alloc] init] : self.toAccountData;
        self.amountData = (self.amountData==nil) ? [[NSMutableArray alloc] init] : self.amountData;
        self.dateData = (self.dateData==nil) ? [[NSMutableArray alloc] init] : self.dateData;
        NSDictionary *input = [NSDictionary dictionaryWithObjectsAndKeys:[Util getKeychain:@"accountNo"],@"accountNo",[Util getKeychain:@"token"],@"token", nil];
        [Http post:@"transferHistory" :input :^(NSDictionary *res)
         {
             if(res==nil)
             {
                 [Util alert:self :@"Connection Problem"];
                 return;
             }
             if([[res[@"message"] lowercaseString] isEqualToString:@"success"])
             {
                 for (id transaction in res[@"transaction"]) {
                     [self.fromAccountData addObject:transaction[@"accountNo"]];
                     [self.toAccountData addObject:transaction[@"toAccountNo"]];
                     [self.amountData addObject:transaction[@"amount"]];
                     [self.dateData addObject:transaction[@"datetime"]];
                 }
                 dispatch_async(dispatch_get_main_queue(), ^{
                     [self.table reloadData];
                 });
             }
             else
             {
                 [Util alert:self :@"An Error Occurred"];
             }
         }];
    });
}


- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [_fromAccountData count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    HistoryCell *c = [tableView dequeueReusableCellWithIdentifier:@"celli"];
    if(c==nil)
    {
        c = [[HistoryCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"celli"];
    }
    c.fromAccount.text = [NSString stringWithFormat:@"%@ : %@",@"From",[_fromAccountData objectAtIndex:[indexPath row]]];
    c.toAccount.text = [NSString stringWithFormat:@"%@ : %@",@"To",[_toAccountData objectAtIndex:[indexPath row]]];
    c.amount.text = [NSString stringWithFormat:@"%@ : %@",@"Amount",[_amountData objectAtIndex:[indexPath row]]];
    c.date.text = [NSString stringWithFormat:@"%@ : %@",@"Date",[_dateData objectAtIndex:[indexPath row]]];
    return c;
}
@end

File: iOS-Swift/watf-bank/PageController/History.swift
import UIKit

class History : UIViewController, UITableViewDelegate, UITableViewDataSource {
    
    @IBOutlet weak var table: UITableView!
    var fromAccountData = Array<String>()
    var toAccountData = Array<String>()
    var amountData = Array<String>()
    var dateData = Array<String>()
    
    override func viewDidLoad() {
        table.delegate = self
        table.dataSource = self
        DispatchQueue.main.async {
            guard let token = KeyChain.load("token"), let acc = KeyChain.load("accountNo") else {
                Util.alert(self, "Invalid Token")
                return
            }
            let input = ["accountNo":acc,"token":token]
            Http().post(input, "transferHistory", completionHandler: {(res: [String:Any]) in
                if let e = res["error"] {
                    Util.alert(self, e as! String)
                }
                else if res["message"] as! String == "Success" {
                    for transaction in res["transaction"] as! Array<Any> {
                        let t = transaction as! [String:String]
                        self.fromAccountData.append(t["accountNo"]!)
                        self.toAccountData.append(t["toAccountNo"]!)
                        self.amountData.append(t["amount"]!)
                        self.dateData.append(t["datetime"]!)
                    }
                    DispatchQueue.main.async {
                        self.table.reloadData()
                    }
                }
                else if let m = res["message"] {
                    Util.alert(self, m as! String)
                }
                else {
                    Util.alert(self, "An Error Occurred")
                }
            })
        }
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return fromAccountData.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var c : HistoryCell
        if let cc = tableView.dequeueReusableCell(withIdentifier: "celli") {
            c = cc as! HistoryCell
        }
        else {
            c = HistoryCell.init(style: .default, reuseIdentifier: "celli")
        }
        c.fromAccount.text = "From : "+fromAccountData[indexPath.section]
        c.toAccount.text = "To : "+toAccountData[indexPath.section]
        c.amount.text = "Amount : "+amountData[indexPath.section]
        c.date.text = "Date : "+dateData[indexPath.section]
        return c
    }
}
